src/couch_mrview/src/couch_mrview_index.erl

```diff
@@ -132,6 +135,7 @@ open(Db, State0) ->
                     % Matching view signatures.
                     NewSt = couch_mrview_util:init_state(Db, Fd, State, Header),
                     ensure_local_purge_doc(Db, NewSt),
+                    check_collator_versions(DbName, NewSt),
```

Comment:
I wonder if we want commit our header upgrade more eagerly here (or as an option perhaps)?

Currently, after we read the header from disk, we update it (in memory!) with the latest collator version, and leave it at that. If there is write eventually on that view, we would write the new header with the now updated list of collator versions. But, if it is just a read-only view then we wouldn't upgrade the header on disk.

It seems, at least for the case when the view is upgraded from the `no-recorded collator version at all` to `record the current collator version` (0 versions to 1 version), we may want to commit in that case as well. I am thinking to have an extra field in the `#mrst` record remembering that we would like to run `commit(NewSt)` operation since we may have upgraded the view.

We do that in the `2.x -> 3.x` upgrade code path since we know we upgrade as the signature changed, I think we should do it in the collator header update case well?

Reply:
I agree with the idea to extend `#mrst{}`. The downside of this approach is that we would need an intermediate release which can handle two versions of `#mrst{}` structure. Because the mrst is sent between nodes and we want to make sure that the cluster would work in a mixed state. The alternative would be to change the return value of `init_state` from

```
init_state(Db, Fd, State, Header) -> #mrst{}.
```

to 

```
init_state(Db, Fd, State, Header) -> {Updated :: boolean(), #mrst{}}.
```

or

```
init_state(Db, Fd, State, Header) -> {OldVsn, NewVsn, #mrst{}}.
```

Reply:
The benefit of returning OldVsn, NewVsn is to be able to have a different logic in open depending on the version bump. I am not sure if it is worth it. Also we can make configuration bump specific. Also not sure if it would be overkill or not.

```ini
[couch_mrview:commit_on_update]

1->2 = true
*->4 = true
```

The exact syntax in the ini config could be different. I just wanted to illustrate the idea.

However we would have to assign explicit version numbers for views formats somehow.

This is definitely would be a huge plus to start including explicit version. The format guessing game was a goto solution for too long.
However it would introduce additional penalty on storage. Also could be out of scope for this PR.

Reply:
I was wrong. I did a quick search and it was too quick. Here https://github.com/apache/couchdb/blob/3.x/src/fabric/src/fabric.erl#L471 we use `#mrst{}`, but we don't send it to another node. I think I mixed up `#mrst{}` and `#mrargs{}`. So we should be ok.

Reply:
@iilyak updated to eagerly commit on upgrade (with a configuration option to disable it if needed). Thanks for the init_state suggestion!